### PR TITLE
is_equal is_not_equal is_eqz is_nez true_ false_

### PR DESCRIPTION
**Pull request by jtlap**

### CODE REVIEW

File: include/eve/module/complex/complex.hpp [hunk 1]
```diff
@@ -116,7 +127,33 @@ namespace eve
     EVE_FORCEINLINE friend  auto  tagged_dispatch(Tag const&, as<Z> const&) noexcept
     {
       detail::callable_object<Tag> cst;
-      return Z{ cst(as<Type>{}), Type{0}};
+      if constexpr(std::same_as<Tag,tag::true__>)
```
Comment (jfalcou) [inline include/eve/module/complex/complex.hpp:130]:
```c++
 if constexpr(std::same_as<Tag,tag::true__> || std::same_as<Tag,tag::false__>)
{
  return cst(as(real(Z{})));
}
else
{
  return Z{ cst(as<Type>{}), Type{0}};
}
 ```

looks better

File: include/eve/module/complex/complex.hpp [hunk 2]
```diff
@@ -116,7 +127,33 @@ namespace eve
...
+      {
+        return Z{ cst(as<Type>{}), Type{0}};
+      }
+    }
+
+    //==============================================================================================
+    // Operator ==,  !=
+    //==============================================================================================
+    template<like<complex> Z1, like<complex> Z2>
+    EVE_FORCEINLINE friend auto operator == (Z1 const& z1, Z2 const& z2) noexcept
```
Comment (jfalcou) [inline include/eve/module/complex/complex.hpp:148]: this is not needed

File: include/eve/module/complex/complex.hpp [hunk 3]
```diff
@@ -116,7 +127,33 @@ namespace eve
...
+    {
+      return (real(z1) == real(z2)) && (imag(z1) == imag(z2));
+    }
+
+    template<like<complex> Z1, like<complex> Z2>
+    EVE_FORCEINLINE friend auto operator != (Z1 const& z1, Z2 const& z2) noexcept
```
Comment (jfalcou) [inline include/eve/module/complex/complex.hpp:154]: this is not needed

File: include/eve/module/complex/regular/detail/arithmetic.hpp [hunk 4]
```diff
@@ -15,7 +15,6 @@
 
 namespace eve::detail
 {
-
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/arithmetic.hpp:18]: noise commit, revert.

File: include/eve/module/complex/regular/detail/predicates.hpp [hunk 5]
```diff
@@ -66,4 +66,37 @@ namespace eve::detail
...
+    return is_ordered(rz, iz);
+  }
+
+  //==============================================================================================
+  //  binary is_equal, is_not_equal
+  //==============================================================================================
+  template<typename Z1,  typename Z2>
+  EVE_FORCEINLINE auto complex_binary_dispatch(eve::tag::is_equal_, Z1 const& z1, Z2 const& z2) noexcept
+  {
+    return (real(z1) == real(z2)) && (imag(z1) == imag(z2));
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/predicates.hpp:82]: this is not needed

File: include/eve/module/complex/regular/detail/predicates.hpp [hunk 6]
```diff
@@ -66,4 +66,37 @@ namespace eve::detail
...
+ }
+
+  template<typename Z1,  typename Z2>
+  EVE_FORCEINLINE auto complex_binary_dispatch(eve::tag::is_not_equal_, Z1 const& z1, Z2 const& z2) noexcept
+  {
+    return (real(z1) != real(z2)) || (imag(z1) != imag(z2));
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/predicates.hpp:88]: this is not needed

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/module/complex/complex.hpp (hunk 1 above)
Comment (jtlap) [reply include/eve/module/complex/complex.hpp:130]: done

File: include/eve/module/complex/complex.hpp (hunk 2 above)
Comment (jtlap) [reply include/eve/module/complex/complex.hpp:148]: suppressed

File: include/eve/module/complex/regular/detail/arithmetic.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/complex/regular/detail/arithmetic.hpp:18]: suppressed

File: include/eve/module/complex/regular/detail/predicates.hpp (hunk 6 above)
Comment (jtlap) [reply include/eve/module/complex/regular/detail/predicates.hpp:88]: suppressed

**Review by jfalcou — APPROVED** [review]

**Pull request merged**